### PR TITLE
Alignment with CTC ASR models powered by k2.

### PR DESCRIPTION
**Pull request by ZhaoZeyu1995**
…els based on k2.

## What does this PR do?

Support alignment with CTC ASR models powered by k2, a Differentiable Weighted Finite-State Transducer toolkit.
Both token- and word-level alignments are supported.

Currently, the training and decoding with k2 for CTC ASR are supported but not alignment.

Dependencies

* k2, for installation, please refer to [the official installation guide](https://k2-fsa.github.io/k2/installation/index.html)

Fixes #<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Comment by ZhaoZeyu1995** [comment]
Here are some examples about how to use the code for alignments. You can also find these examples in the comment/documentation in the code.

```Python
    >>> import torch
    >>> from speechbrain.pretrained import EncoderASR
    >>> from speechbrain.k2_integration.align import CTCAligner
    >>> asr_model = EncoderASR.from_hparams(source="speechbrain/asr-wav2vec2-librispeech", savedir="pretrained_models/asr-wav2vec2-librispeech")
    >>> device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    >>> aligner = CTCAligner(model=asr_model, tokenizer=asr_model.tokenizer, device=device)
    >>> audio_files = ["samples/audio_samples/example1.wav", "samples/audio_samples/example2.wav"]
    >>> transcripts = ["HELLO WORLD", "THIS IS SPEECHBRAIN"]
    >>> # align one audio file to tokens
    >>> alignment = aligner.align_audio_to_tokens(audio_files[0], transcripts[0])
    >>> # align one audio file to words
    >>> alignment = aligner.align_audio_to_words(audio_files[0], transcripts[0], frame_shift=0.02)
    >>> # align a batch of audio files to tokens
    >>> alignments = aligner.align_batch_to_tokens(audio_files, transcripts)
    >>> # align a batch of audio files to words
    >>> alignments = aligner.align_batch_to_words(audio_files, transcripts, frame_shift=0.02)
    >>> # align a csv file to tokens
    >>> aligner.align_csv_to_tokens("samples/audio_samples/example.csv", "samples/audio_samples/example_token_alignment.txt")
    >>> align a csv file to words
    >>> # aligner.align_csv_to_words("samples/audio_samples/example.csv", "samples/audio_samples/example_word_alignment.csv", frame_shift=0.02)
```

**Comment by ZhaoZeyu1995** [comment]
Here are some examples of input and output files.

Input: dev-clean.csv (from LibriSpeech)

[dev-clean.csv](https://github.com/user-attachments/files/17921433/dev-clean.csv)

Outputs: dev-clean.token.ali.txt

[dev-clean.token.ali.txt](https://github.com/user-attachments/files/17921448/dev-clean.token.ali.txt)

Outputs: dev-clean.word.ali.csv

[dev-clean.word.ali.csv](https://github.com/user-attachments/files/17921451/dev-clean.word.ali.csv)

**Comment by ZhaoZeyu1995** [comment]
No problem. I will deal with them soon.

**Comment by ZhaoZeyu1995** [comment]
Hello @mravanelli , I just solved the pre-commit issue.

**Comment by TParcollet** [comment]
@pplantinga I see this as a sort of integration. Should this be relocated in your integration folder?

File: speechbrain/k2_integration/align.py [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+        log_prob_len: torch.Tensor,
+        targets: List[List[int]],
+    ) -> List[List[int]]:
+        """
+        Align targets to log_probs.
+
+        Arguments
+        ---------
+        log_prob: torch.Tensor
+            A tensor of shape (N, T, C) containing the log-probabilities.
```
Comment (pplantinga) [inline speechbrain/k2_integration/align.py]: I'm not sure if everyone will know what N, T, and C refer to. Maybe this could mention the meanings of "batch", "time", "vocabulary" to be more explicit?

File: speechbrain/k2_integration/align.py [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+        """
+        # Basic checks.
+        assert log_prob.ndim == 3
+        assert log_prob_len.ndim == 1
+        assert log_prob.shape[0] == log_prob_len.shape[0]
+        assert isinstance(targets, list)
+        assert isinstance(targets[0], list)
+        assert log_prob.shape[0] == len(targets)
+
+        N, T, C = log_prob.shape
```
Comment (pplantinga) [inline speechbrain/k2_integration/align.py]: Is this line needed?

File: speechbrain/k2_integration/align.py [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+        df = pd.read_csv(input_csv)
+        audio_files = df["wav"].tolist()
+        transcripts = df["wrd"].tolist()
+        ids = df["ID"].tolist()
+
+        if frame_shift is None or frame_shift == 1:
+            logger.info("No frame shift is provided or the frame shift is 1.")
+            logger.info("The resulting alignment will be in frame index.")
+            logger.info("The frame index starts from 0.")
+            frame_shift = 1
```
Comment (pplantinga) [inline speechbrain/k2_integration/align.py]: I'm not sure why frame_shift == 1 is treated specially here, is this the value one should pass to disable?

File: speechbrain/k2_integration/align.py [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+except ImportError:
+    MSG = "Cannot import k2, so training and decoding with k2 will not work.\n"
+    MSG += "Please refer to https://k2-fsa.github.io/k2/installation/from_wheels.html for installation.\n"
+    MSG += "You may also find the precompiled wheels for your platform at https://download.pytorch.org/whl/torch_stable.html"
+    raise ImportError(MSG)
+
+
+class Aligner(abc.ABC):
+    """
+    Abstract class for aligner.
```
Comment (pplantinga) [inline speechbrain/k2_integration/align.py]: I'm not 100% sold on the idea of having an abstract class and concrete class here rather than just one unified class `CTCAligner` -- I can see that some uses may have different ways to extract text tokens or log probabilities, but these don't necessarily require defining two different classes here. You could just define one class that works for most use cases and a user could sub-class and override the tensor extraction portions.

File: speechbrain/integrations/k2_fsa/align.py [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+
+    Example
+    -------
+    >>> import torch
+    >>> from speechbrain.inference import EncoderASR
+    >>> from speechbrain.integrations.k2_fsa.align import CTCAligner
+    >>> asr_model = EncoderASR.from_hparams(source="speechbrain/asr-wav2vec2-librispeech", savedir="pretrained_models/asr-wav2vec2-librispeech")
+    >>> device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    >>> aligner = CTCAligner(model=asr_model, tokenizer=asr_model.tokenizer, device=device)
+    >>> audio_files = ["samples/audio_samples/example1.wav", "samples/audio_samples/example2.wav"]
```
Comment (pplantinga) [inline speechbrain/integrations/k2_fsa/align.py]: These paths don't seem to exist anymore, can you update?

**Review by pplantinga — APPROVED** [review]
I was able to confirm that this works. I have updated the branch and it looks read to merge for me
